pkg/controller/volume/attachdetach/metrics/metrics.go

```diff
@@ -0,0 +1,161 @@
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+var (
+	inUsePVCMetricDesc = prometheus.NewDesc(
+		prometheus.BuildFQName("", "storage_counter", "pvc_in_use"),
+		"Measure number of pvcs in use",
+		[]string{"node", "volume_plugin"}, nil)
+)
+var registerMetrics sync.Once
+
+type pvcInUseCollector struct {
+	pvcLister       corelisters.PersistentVolumeClaimLister
+	podLister       corelisters.PodLister
+	pvLister        corelisters.PersistentVolumeLister
+	volumePluginMgr *volume.VolumePluginMgr
+}
+
+// nodePVCCount contains map of {"nodeName": {"pluginName": pvc_count }}
+// For example :
+//     node 172.168.1.100.ec2.internal has 10 EBS and 3 glusterfs PVC in use
+//     {"172.168.1.100.ec2.internal": {"aws-ebs": 10, "glusterfs": 3}}
+type nodePVCCount map[types.NodeName]map[string]int
+
+// Register registers pvc's in-use metrics
+func Register(pvcLister corelisters.PersistentVolumeClaimLister,
+	pvLister corelisters.PersistentVolumeLister,
+	podLister corelisters.PodLister,
+	pluginMgr *volume.VolumePluginMgr) {
+	registerMetrics.Do(func() {
+		prometheus.MustRegister(newPVCInUseCollector(pvcLister, podLister, pvLister, pluginMgr))
+	})
+
+}
+
+func (pvcInUse nodePVCCount) add(nodeName types.NodeName, pluginName string) {
+	nodeCount, ok := pvcInUse[nodeName]
+	if !ok {
+		nodeCount = map[string]int{}
+	}
+	nodeCount[pluginName]++
+	pvcInUse[nodeName] = nodeCount
+}
+
+func newPVCInUseCollector(
+	pvcLister corelisters.PersistentVolumeClaimLister,
+	podLister corelisters.PodLister,
+	pvLister corelisters.PersistentVolumeLister,
+	pluginMgr *volume.VolumePluginMgr) *pvcInUseCollector {
+	return &pvcInUseCollector{pvcLister, podLister, pvLister, pluginMgr}
+}
+
+func (collector *pvcInUseCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- inUsePVCMetricDesc
+}
+
+func (collector *pvcInUseCollector) Collect(ch chan<- prometheus.Metric) {
+	nodePVCMap := collector.getPVCUseByNode()
+	for nodeName, pluginCount := range nodePVCMap {
+		for pluginName, count := range pluginCount {
+			metric, err := prometheus.NewConstMetric(inUsePVCMetricDesc,
+				prometheus.GaugeValue,
+				float64(count),
+				string(nodeName),
+				pluginName)
+			if err != nil {
+				glog.Warningf("Failed to create metric : %v", err)
+			}
+			ch <- metric
+		}
+	}
+}
+
+func (collector *pvcInUseCollector) getPVCUseByNode() nodePVCCount {
+	pods, err := collector.podLister.List(labels.Everything())
+	if err != nil {
+		glog.Errorf("Error getting pod list")
+		return nil
+	}
+
+	nodePVCMap := make(nodePVCCount)
+	for _, pod := range pods {
+		if len(pod.Spec.Volumes) <= 0 {
+			continue
+		}
+
+		nodeName := types.NodeName(pod.Spec.NodeName)
+		if nodeName == "" {
+			continue
+		}
+		for _, podVolume := range pod.Spec.Volumes {
```

Comment:
So this is actually just counting pods that have been scheduled to a node, but not necessarily attached. Is that what we want?